### PR TITLE
DISTX-186. Improve user sync behavior

### PR DESCRIPTION
**Pull request by handavid**
User sync behavior has changed in the following ways:
  - All groups are retrieved from UMS instead of only those that
    users are members of
  - All users are added to IPA instead of only those that are members
    of a group
  - Users are retrieved a single time from the UMS instead of
    repeatedly for each environment
  - A user filter was implemented to allow sync of specified users
  - Group membership is calculated and compared between UMS and IPA
  - Users are added to groups in IPA

Other minor changes were:
  - account and actor crn are retrieved in the controller layer
    and passed in to the UserService rather than retrieved in
    the UmsUsersStateProvider

This commit does not remove or deactivate users, nor does it
remove users from groups.

### CODE REVIEW

**Comment by keyki** [comment]
Do we need to extend the ums mock with new api calls?

**Comment by handavid** [comment]
> Do we need to extend the ums mock with new api calls?

I'll look into it. We will need a new listGroups call

**Comment by keyki** [comment]
Right now as we auto sync users when an IPA cluster is created we extended the mock https://github.com/hortonworks/cloudbreak/blob/master/mock-caas/src/main/java/com/sequenceiq/caas/grpc/service/MockUserManagementService.java to support this scenario. If we can keep extending it even with dummy data just to make it work that would be good. 

If you want to run this mock from idea you need to extend the Profile with the CB_LOCAL_DEV_LIST with `auth-mock,caas-api`

File: freeipa/src/main/java/com/sequenceiq/freeipa/controller/UserV1Controller.java [hunk 1]
```diff
@@ -42,13 +42,16 @@
     public SynchronizeUserResponse synchronizeUser(SynchronizeUserRequest request) {
         String userCrn = checkUserCrn();
         LOGGER.debug("synchronizeUser() requested for user {}", userCrn);
+        String accountId = crnService.getCurrentAccountId();
 
-        return userService.synchronizeUser(userCrn);
+        return userService.synchronizeUser(accountId, userCrn, userCrn);
```
Comment (abhishekaypurohit) [inline freeipa/src/main/java/com/sequenceiq/freeipa/controller/UserV1Controller.java]: @handavid - nit - userCrn is repeating here.

Comment (handavid) [reply freeipa/src/main/java/com/sequenceiq/freeipa/controller/UserV1Controller.java]: this is on purpose. the userCrn is passed in first as the actor and secondly as the subject.

**Comment by handavid** [comment]
@keyki I updated the mock ums and ran a sync.

File: freeipa/src/main/java/com/sequenceiq/freeipa/controller/UserV1Controller.java (hunk 1 above)
Comment (abhishekaypurohit) [reply freeipa/src/main/java/com/sequenceiq/freeipa/controller/UserV1Controller.java]: ok sure, please add comments for the same in the code in later PRs.

**Pull request merged**